File: Kamynin/1391.cpp
class Solution {
public:
    bool hasValidPath(vector<vector<int>>& grid) {
	int m = grid.size(), n = grid[0].size();
    if (grid[0][0] == 5 || grid[m-1][n-1] == 4)
		return false;
	int i = 0, j = 0, d = -1, c = 0;
    int startDir = grid[0][0] == 4 ? 3 : -1;
    bool isCorrect = true;
	switch (grid[0][0])
	{
		case 1: d = 4; break;
		case 2: d = 1; break;
        case 3: d = 4; break;
		case 4: d = startDir; break;
		case 6: d = 1; break;
		default: return false;
	}
	while (true)
	{
        if (!isCorrect){
            if (startDir == 3)
                i = 0, j = 0, d = 2, c = 0, startDir = 2;
            else
                return false;
        }
        if (i == m - 1 && j == n - 1){
        	int v = grid[i][j];
            if (d == 1 && (v == 2 || v == 5 || v == 6) ||
                d == 2 && (v == 1 || v == 4 || v == 6) ||
                d == 3 && (v == 2 || v == 3 || v == 4) ||
                d == 4 && (v == 1 || v == 3 || v == 5))
                return true;
        }
		
		if ((c > m * n || i < 0 || j < 0 || i >= m || j >= n) && startDir == 3)
			i = 0, j = 0, d = 2, c = 0, startDir = 2;   
	    else if (i < 0 || j < 0 || i >= m || j >= n)
            break;
        else if (c > m * n)
			break;
        if (grid[i][j] == 1) {
            if (d == 2 || d == 4) d == 2 ? --j : ++j;
            else isCorrect = false; 
        }
        else if (grid[i][j] == 2) {
            if (d == 1 || d == 3) d == 1 ? ++i : --i;
            else isCorrect = false; 
        }
        else if (grid[i][j] == 3){
            if (d == 3 || d == 4) d == 3 ? (--j, d = 2) : (++i, d = 1);
            else isCorrect = false; 
        }
        else if (grid[i][j] == 4) {
            if (d == 2 || d == 3) d == 2 ? (++i, d = 1) : (++j, d = 4); 
            else isCorrect = false; 
        }
        else if (grid[i][j] == 5) {
            if (d == 1 || d == 4) d == 1 ? (--j, d = 2) : (--i, d = 3); 
            else isCorrect = false; 
        }
        else if (grid[i][j] == 6) {
            if (d == 1 || d == 2) d == 1 ? (++j, d = 4) : (--i, d = 3);
            else isCorrect = false;
        } 
        ++c; 
        
	}
	return false;
}
};

File: Бабат/1391.py
class Solution:
    def hasValidPath(self, grid: List[List[int]]) -> bool:
        def num(A,me,m):
            if (A==1):
                q=me-1
                w=me+1
            if (A==2):
                q=me-m
                w=me+m
            if (A==3):
                q=me-1
                w=me+m
            if (A==4):
                q=me+1
                w=me+m
            if (A==5):
                q=me-m
                w=me-1
            if (A==6):
                q=me-m
                w=me+1
            return(q,w)
        flag=0
        A=grid
        n=len(A)
        m=len(A[0])
        me=0
        last=0
        while(me!=(n)*(m)-1):
            i=me//m
            j=me%m
            q,w=num(A[i][j],me,m)
            if me==0:
                if (q>0)and(w<=0)and(q<n*m):
                    last=me
                    me=q
                elif (w>0)and(q<=0)and(w<n*m):
                    last=me
                    me=w
                elif (flag==1):
                    last=me
                    me=q
                    flag=0
                elif (w>0)and(q>0):
                    if (w>=n*m):
                        last=me
                        me=q
                    elif (q>=n*m):
                        last=me
                        me=w
                    else:
                        last=me
                        me=w
                        flag=1
                else: return(False)
            elif (q==last):
                if(w<=0)or((i!=w//m)and(j!=w%m))or(w>=n*m): 
                    if flag==0: return(False)
                    last=0
                    me=0
                last=me
                me=w
            elif (w==last):
                if(q<=0)or((i!=q//m)and(j!=q%m))or(q>=n*m):
                    if flag==0: return(False)
                    last=0
                    me=0
                last=me
                me=q
            else: 
                if flag==0: return(False)
                last=0
                me=0
            if (me==(n)*(m)-1):
                q,w=num(A[n-1][m-1],me,m)
                if (q!=last)and(w!=last):
                    if flag==0: return(False)
                    last=0
                    me=0
        return(True)
